Segfault - nil pointer dereference

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x7ed6b6]

goroutine 17231123 [running]:
ktbs.dev/mubeng/pkg/mubeng.(*Proxy).New(0xc0316e6f50, 0xc146b1c600)
        /root/mubeng/pkg/mubeng/mubeng.go:27 +0x156
ktbs.dev/mubeng/internal/server.(*Proxy).onRequest.func1(0xc146b1c600)
        /root/mubeng/internal/server/handler.go:63 +0x2cd
created by ktbs.dev/mubeng/internal/server.(*Proxy).onRequest
        /root/mubeng/internal/server/handler.go:44 +0x372
```

Unfortunately I'm not able to reproduce it directly but it already happened a few times (running on ubuntu, latest master version, go 1.20), after pushing a few hundred thousand requests through the proxy server at a rate of a few hundreds per second.

Another one, not sure it's related but this time it logged an error right before a segfault
```
2024/06/03 21:12:36 [ERRO] ▶ [::1]:33684 parse "http://user:pass@12.34.567.89:12345\n": net/url: invalid control character in URL
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x7ed6b6]

goroutine 5356869 [running]:
ktbs.dev/mubeng/pkg/mubeng.(*Proxy).New(0xc03051cf50, 0xc009046600)
        /root/mubeng/pkg/mubeng/mubeng.go:27 +0x156
ktbs.dev/mubeng/internal/server.(*Proxy).onRequest.func1(0xc009046600)
        /root/mubeng/internal/server/handler.go:63 +0x2cd
created by ktbs.dev/mubeng/internal/server.(*Proxy).onRequest
        /root/mubeng/internal/server/handler.go:44 +0x372
```
I thought it's the last entry in the proxies file list but no, it's somewhere in the middle.

--- a/internal/checker/checker.go
+++ b/internal/checker/checker.go
@@ -78,7 +78,11 @@ func check(address string, timeout time.Duration) (IPInfo, error) {
 		Transport: tr,
 	}
 
-	client, req = proxy.New(req)
+	client, err := proxy.New(req)
+	if err != nil {
+		return ipinfo, err
+	}
+
 	client.Timeout = timeout
 	req.Header.Add("Connection", "close")
 

--- a/internal/checker/vars.go
+++ b/internal/checker/vars.go
@@ -1,11 +1,6 @@
 package checker
 
-import (
-	"net/http"
-)
-
 var (
-	client *http.Client
 	ipinfo IPInfo
 
 	endpoint = "https://ipinfo.io/json"

--- a/internal/server/handler.go
+++ b/internal/server/handler.go
@@ -60,7 +60,12 @@ func (p *Proxy) onRequest(req *http.Request, ctx *goproxy.ProxyCtx) (*http.Reque
 			Transport: tr,
 		}
 
-		client, req = proxy.New(req)
+		client, err := proxy.New(req)
+		if err != nil {
+			resChan <- err
+			return
+		}
+
 		client.Timeout = p.Options.Timeout
 		if p.Options.Verbose {
 			client.Transport = dump.RoundTripper(tr)

--- a/internal/server/server.go
+++ b/internal/server/server.go
@@ -16,13 +16,19 @@ import (
 // An active log have 2 receivers, especially stdout and into file if opt.Output isn't empty.
 // Then close the proxy server if it receives a signal that interrupts the program.
 func Run(opt *common.Options) {
+	var recs []*logo.Receiver
+
 	cli := logo.NewReceiver(os.Stderr, "")
 	cli.Color = true
 	cli.Level = logo.DEBUG
+	recs = append(recs, cli)
 
-	file, _ := logo.Open(opt.Output)
-	out := logo.NewReceiver(file, "")
-	out.Format = "%s: %s"
+	file, err := logo.Open(opt.Output)
+	if err == nil {
+		out := logo.NewReceiver(file, "")
+		out.Format = "%s: %s"
+		recs = append(recs, out)
+	}
 
 	dump = &httpretty.Logger{
 		RequestHeader:  true,
@@ -43,7 +49,7 @@ func Run(opt *common.Options) {
 		Handler: handler.HTTPProxy,
 	}
 
-	log = logo.NewLogger(cli, out)
+	log = logo.NewLogger(recs...)
 
 	if opt.Watch {
 		watcher, err := opt.ProxyManager.Watch()

--- a/internal/server/vars.go
+++ b/internal/server/vars.go
@@ -12,7 +12,6 @@ var (
 	rotate  string
 	handler *Proxy
 	server  *http.Server
-	client  *http.Client
 	dump    *httpretty.Logger
 	mime    = "text/plain"
 	log     *logo.Logger

--- a/pkg/mubeng/mubeng.go
+++ b/pkg/mubeng/mubeng.go
@@ -11,7 +11,7 @@ import (
 //
 // also removes Hop-by-hop headers when it is sent to backend (see http://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html),
 // then add X-Forwarded-For header value with the IP address value of rotator proxy IP.
-func (proxy *Proxy) New(req *http.Request) (*http.Client, *http.Request) {
+func (proxy *Proxy) New(req *http.Request) (*http.Client, error) {
 	client = &http.Client{Transport: proxy.Transport}
 
 	// http: Request.RequestURI can't be set in client requests.
@@ -22,7 +22,10 @@ func (proxy *Proxy) New(req *http.Request) (*http.Client, *http.Request) {
 		req.Header.Del(h)
 	}
 
-	proxyURL, _ := url.Parse(proxy.Address)
+	proxyURL, err := url.Parse(proxy.Address)
+	if err != nil {
+		return client, err
+	}
 
 	if host, _, err := net.SplitHostPort(proxyURL.Host); err == nil {
 		if prior, ok := req.Header["X-Forwarded-For"]; ok {
@@ -33,5 +36,5 @@ func (proxy *Proxy) New(req *http.Request) (*http.Client, *http.Request) {
 
 	req.Header.Set("X-Forwarded-Proto", req.URL.Scheme)
 
-	return client, req
+	return client, nil
 }

--- a/pkg/mubeng/mubeng_test.go
+++ b/pkg/mubeng/mubeng_test.go
@@ -39,7 +39,7 @@ func TestProxyNew(t *testing.T) {
 		fields fields
 		args   args
 		want   *http.Client
-		want1  *http.Request
+		want1  error
 	}{
 		{
 			name: "New Proxy",
@@ -53,7 +53,7 @@ func TestProxyNew(t *testing.T) {
 			want: &http.Client{
 				Transport: tr,
 			},
-			want1: req,
+			want1: nil,
 		},
 	}
 

